### ISSUE TITLE
How can we calculate the loss for vima?

### ISSUE THREAD

**Issue by Mano176** [vimalabs/VIMA#28]
Hi there,

I am trying to train the VIMA model on a custom dataset to generate custom actions. I am a bit confused though on how to determine the loss while training. 
My first attempt was to generate action tokens for the actions from my dataset like this `action_tokens = policy.forward_action_token(actions)` and predicted action tokens like this `predicted_action_tokens = policy.forward(...)` and then just forward these into the cross entropy loss function. But I realized you don't have to pass the tokens but rather some logits. But I can't figure out how to access these.
My guess would be to somehow extract them from `dist_dict = policy.forward_action_decoder(predicted_action_tokens)`.

Also the actions I want to generate are partly categorical and partly multicategorical and I took from [this issue](https://github.com/vimalabs/VIMA/issues/11) that for multicategorical actions you didn't even used cross entropy loss but rather either some type of regression or GMM (which is gaussian mixture modelling?). If that's correct I hope you could give my some insight on what exactly you applied.

Like the person in [this issue](https://github.com/vimalabs/VIMA/issues/25) already said, a pseudo-code of the training algorithm would also be really helpful to reproduce and build upon your results. :)

Best regards,
Mano

**Comment by yunfanjiang** [comment]
Hi there, thank you for your interest in our project. To answer your questions:

This snippet would illustrate how loss is calcualted.
```
# compute prompt tokens
prompt_tokens, prompt_masks = policy.forward_prompt_assembly(prompts)

# compute obs tokens
obs_tokens, obs_masks = policy.forward_obs_token(obs)

# compute action tokens
# discretize if needed
action = self.discretize_action(action)
# cache target action
tar_action = {k: v.clone() for k, v in action.items()}
# slice action sequence up to the last one
action = slice(action, np.s_[:-1])
action_tokens = policy.forward_action_token(action)

# get predicted action tokens
pred_action_tokens = policy.forward(
    obs_token=obs_tokens,
    action_token=action_tokens,
    prompt_token=prompt_tokens,
    prompt_token_mask=prompt_masks,
    obs_mask=obs_masks,
)

# forward action decoder
dist_dict = policy.forward_action_decoder(pred_action_tokens)

# compute imitation loss
imitation_loss = dist_dict.imitation_loss(actions=tar_action)
```
The `imitation_loss` method internally implements `torch.nn.functional.cross_entropy` for loss calculation. 

> for multicategorical actions you didn't even used cross entropy loss but rather either some type of regression or GMM

We indeed used cross-entropy loss to optimize (multi-)categorical actions. We mentioned that we experimented with GMM (Gaussian Mixture Model) to model actions in continuous format, but eventually opted for discrete actions.

**Comment by amitkparekh** [comment]
Where does the `imitation_loss` method come from? 

It is not found within this repo, nor is it from torch or similar. Since `dist_dict` is a dictionary (explicitly, it's `dict[Literal['pose0_position', 'pose0_rotation', 'pose1_position', 'pose1_rotation'], MultiCategorical]`), it doesn't have a `imitation_loss` method. 

If I'm missing something, can you point me in the right direction?
